File: lib/gsd/task.rb
module GSD
  class Task
    attr_accessor :id, :title, :description, :deadline, :done

    def self.all
      DB.instance.tasks.map do |attributes|
        new(
          id: attributes[0],
          title: attributes[1],
          description: attributes[2],
          deadline: attributes[3],
          done: attributes[4]
        )
      end
    end

    def self.find(id)
      attributes = DB.instance.find_task(id)
      new(
        id: attributes[0],
        title: attributes[1],
        description: attributes[2],
        deadline: attributes[3],
        done: attributes[4]
      )
    end

    def initialize(id: nil, title: nil, description: nil, deadline: nil, done: false)
      @id = id
      @title = title
      @description = description
      @deadline = deadline
      @done = done
    end

    def done?
      done == 1
    end

    def late?
      !done? && (Date.parse(deadline) < Date.today)
    end

    def save
      if !title.nil? && title != "" && !deadline.nil?
        self.id = DB.instance.create_task(
          title, description, deadline, done
        )
        true
      else
        false
      end
    end

    def done!
      DB.instance.task_done(id)
      self.done = true
    end

    def delete
      DB.instance.delete_task(id)
    end
  end
end

File: src/task.js
const db = require("./db");

class Task {
  static all(callback) {
    db.tasks( (rows) => {
      const tasks = rows.map(r => new Task(r));
      callback(tasks);
    });
  }

  static find(id, callback) {
    db.findTask(id, (row) => {
      const foundTask = new Task(row);
      callback(foundTask);
    });
  }

  constructor(params) {
    this.id = params.id;
    this.title = params.title;
    this.description = params.description;
    this.deadline = new Date(params.deadline);
    this.done = params.done;
  }

  late() {
    return (!this.done && (new Date(this.deadline) < new Date()));
  }

  save() {
    if (this.title && this.title != "" && this.deadline) {
      db.createTask(this.title, this.description, this.deadline, this.done, (id) => {
        this.id = id;
      });
      return true;
    }
    return false;
  }

  markDone() {
    db.taskDone(this.id);
    this.done = true;
  }

  destroy() {
    db.deleteTask(this.id);
  }
}

module.exports = Task
